upgrades/upgrade_testing_framework/clients/rest_client_default.py

```diff
@@ -0,0 +1,105 @@
+import json
+
+from upgrade_testing_framework.clients.rest_client_base import RESTClientBase
+import upgrade_testing_framework.clients.rest_ops as ops
+
+class RESTClientDefault(RESTClientBase):
+    """
+    Default REST client for interfacing w/ Elasticsearch and OpenSearch clusters.
+
+    Given our current limited knowledge of where REST incompatibilities will pop up between versions, we will
+    arbitrarily assume that the interface that OpenSearch 1.3.6 presents is our baseline that all other versions will
+    conform to/deviate from.  If this later turns out to be a poor assumption, we can iterate on this code.  We'll know
+    we need to re-examine this assumption if we end up with a large number of version-specific client implementations.
+
+    You might ask, "why make a rest client at all instead of using one of the client SDKs?"  The answer to that is
+    we're trying to avoid introducing another source of bugs/incompatibility (language-specific SDKs) on top of those
+    found in the core engine's REST interface.  Additionally, we believe that many customers write their own clients
+    against the REST interface, so writing one for the UTF mimics their use-case more closely.  As always, additional
+    experience and feedback may dictate we pivot in the future.
+    """
+
+    def get_node_info(self, port: int) -> ops.RESTResponse:
+        """
+        Get basic about the specific node you hit, such as version.
+        """
+        rest_path = ops.RESTPath(port = port)
+        return ops.perform_get(rest_path = rest_path)
+
+    def get_nodes_status(self, port: int) -> ops.RESTResponse:
+        """
+        Get info on the cluster's nodes.
+        """
+        rest_path = ops.RESTPath(port = port, suffix = "_cat/nodes")
+        params = {"v": "true", "pretty": "true"}
+        return ops.perform_get(rest_path = rest_path, params = params)
+
+    def get_doc_by_id(self, port: int, index: str, doc_id: str) -> ops.RESTResponse:
+        """
+        Get a document by its ID
+        """
+        rest_path = ops.RESTPath(port = port, suffix = f"{index}/_doc/{doc_id}")
+        params = {"pretty": "true"}
+        return ops.perform_get(rest_path = rest_path, params = params)
+
+    def post_doc_to_index(self, port: int, index: str, doc: dict) -> ops.RESTResponse:
+        """
+        Post a single document to an index
+        """
+        rest_path = ops.RESTPath(port = port, suffix = f"{index}/_doc")
+        params = {"pretty": "true"}
+        headers={"Content-Type": "application/json"}
+
+        return ops.perform_post(rest_path = rest_path, data = json.dumps(doc), headers = headers, params = params)
+
+    def create_snapshot(self, port: int, repo: str, snapshot_id: str) -> ops.RESTResponse:
+        """
+        Create a snapshot of the cluster into a repo
```

Comment:
It wasn't very clear that `create_snapshot` here is blocking but `restore_snapshot` isn't. Maybe add "Create a snapshot of the cluster into a repo, and wait for its completion"

Reply:
Good callout, will address

Reply:
For now, I'm just going to make both of these commands non-blocking.

Reply:
I'm okay with this blocking, just wanted to make sure it was consistent/clear. In our next change we should add a temporary sleep() command for create snapshot (similar to restore) to give enough time for the snapshot to complete until we get our actual logic in

Reply:
I think I'd prefer to actually check for the right state actively and wait for it that way.  The sleep is good reminder for us to do something better, as we all agree it's a hack.